src/array.jl

```diff
@@ -738,3 +738,82 @@ summary(A::CategoricalArray{T, N, R}) where {T, N, R} =
 
 refs(A::CategoricalArray) = A.refs
 pool(A::CategoricalArray) = A.pool
+
+"""
+    sort!(V::CategoricalVector; rev::Bool=false)
+
+A fast in-place sorting method for `CategoricalVector`` using counting sort
+""" 
+function is_categorical_after_by(by::Function, v::CategoricalVector)
+    for v1 in v
+        if !ismissing(v1) 
+            return isa(by(v1), CatValue) 
+        elseif isa(by(v1), CatValue) # if v1 is missing but `by`` still converts it to a `CatValue`
+            return true
+        end
+    end
+    # if all values were missing then it will reach here
+    return false
+end
+
+function Base.sort!(v::CategoricalVector;
+                    alg::Base.Algorithm=Base.Sort.defalg(v),
+                    lt=isless,
+                    by=identity,
+                    rev::Bool=false,
+                    order::Base.Ordering=Base.Forward)
+
+    ordr = Base.ord(lt,by,rev,order)
+    if (ordr == Base.Forward || ordr == Base.Reverse) && is_categorical_after_by(by, v)
+        if ordr == Base.Forward
+            rev = false
+        else ordr == Base.Reverse
+            rev = true
+        end
+            
+        # add one more level to `cnts` in case there is missing
+        cnts = zeros(UInt, length(v.pool) + 1)
+
+        # do a count/histogram of the references, `v.refs`;
+        # please note that `v.refs`` may not be in the right order, i.e. ref = 1
+        # does not mean that it should be ranked first. The right order is 
+        # determined by `v.pool.order``
+        @inbounds for ref in v.refs
+            cnts[ref + 1] += 1
+        end
+
+        # compute the order in which to read from `cnts``
+        # based on the correct ordering in `x.pool.order`
+        j = 0
+        if eltype(v) >: Missing
+            missing_cnt = cnts[1]
+            if rev == false
```

Comment:
Better swap the two branches to get rid of that `== false`.

Reply:
done